include/ginkgo/core/base/scoped_device_id.hpp

```diff
@@ -0,0 +1,229 @@
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_PUBLIC_CORE_BASE_SCOPED_DEVICE_ID_HPP_
+#define GKO_PUBLIC_CORE_BASE_SCOPED_DEVICE_ID_HPP_
+
+
+#include <memory>
+
+
+namespace gko {
+
+
+class OmpExecutor;
+class CudaExecutor;
+class HipExecutor;
+class DpcppExecutor;
+
+
+namespace detail {
+
+
+/**
+ * A RAII, move-only base class for the scoped device id used for different
+ * executors.
+ */
+class generic_scoped_device_id {
+public:
+    generic_scoped_device_id() = default;
+
+    // TODO: this should be a purely virtual funtion, but somehow that leads to
+    // linker errors
+    // This is explicitly not noexcept, since setting the device id may throw in
+    // the derived classes for CUDA and HIP.
+    virtual ~generic_scoped_device_id() noexcept(false){};
+
+    // Prohibit copy construction
+    generic_scoped_device_id(const generic_scoped_device_id& other) = delete;
+
+    // Prohibit copy assignment
+    generic_scoped_device_id& operator=(const generic_scoped_device_id& other) =
+        delete;
+};
+
+
+/**
+ * An implementation of generic_scoped_device_id that does nothing.
+ *
+ * This is used for OmpExecutor and DpcppExecutor, since they don't require
+ * setting a device id.
+ */
+class noop_scoped_device_id : public generic_scoped_device_id {
+public:
+    noop_scoped_device_id() = default;
+    ~noop_scoped_device_id() noexcept(false) override {}
+};
+
+
+/**
+ * A scoped device id for CUDA.
+ */
+class cuda_scoped_device_id : public generic_scoped_device_id {
+public:
+    explicit cuda_scoped_device_id(int device_id);
+
+    ~cuda_scoped_device_id() noexcept(false) override;
+
+    cuda_scoped_device_id(cuda_scoped_device_id&& other) noexcept;
+
+    cuda_scoped_device_id& operator=(cuda_scoped_device_id&& other) noexcept;
```

Comment:
What I'm missing for now is what happens when not compiling `CUDA`. These methods are never defined then, no? Shouldn't they be to at least `GKO_NOT_COMPILED` like for executor functions? Since users could be using these anywhere since they are publicly accessible.

Same for HIP.

Reply:
I added these to the device_hooks. I think that should be enough